File: backend/app/repositories/roles.py
import uuid

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.models import Opportunity, Role


class RoleRepository:
    def __init__(self, db: Session) -> None:
        self.db = db

    def create(
        self,
        *,
        user_id: uuid.UUID,
        workspace_id: uuid.UUID,
        company_id: uuid.UUID,
        source_id: uuid.UUID | None,
        title: str,
        job_url: str | None = None,
        location: str | None = None,
        remote_type: str | None = None,
        compensation_min=None,
        compensation_max=None,
        compensation_currency: str | None = None,
        raw_description: str | None = None,
        normalized_description: str | None = None,
        parse_metadata: dict | None = None,
        status: str = "found",
        date_found=None,
        date_posted=None,
    ) -> Role:
        role = Role(
            user_id=user_id,
            workspace_id=workspace_id,
            company_id=company_id,
            source_id=source_id,
            title=title,
            job_url=job_url,
            location=location,
            remote_type=remote_type,
            compensation_min=compensation_min,
            compensation_max=compensation_max,
            compensation_currency=compensation_currency,
            raw_description=raw_description,
            normalized_description=normalized_description,
            parse_metadata=parse_metadata or {},
            status=status,
            date_found=date_found,
            date_posted=date_posted,
        )
        self.db.add(role)
        self.db.flush()
        self.db.refresh(role)
        return role

    def get_by_id_for_user(
        self,
        *,
        role_id: uuid.UUID,
        user_id: uuid.UUID,
    ) -> Role | None:
        statement = select(Role).where(
            Role.id == role_id,
            Role.user_id == user_id,
            Role.deleted_at.is_(None),
        )
        return self.db.scalar(statement)

    def list_for_user(self, *, user_id: uuid.UUID) -> list[Role]:
        statement = (
            select(Role)
            .where(
                Role.user_id == user_id,
                Role.deleted_at.is_(None),
                Role.status != "archived",
            )
            .order_by(Role.date_found.desc(), Role.created_at.desc())
        )
        return list(self.db.scalars(statement))


OpportunityRepository = RoleRepository

File: frontend/src/types/roles.ts
export type RoleStatus = "found" | "interested" | "applied" | "archived";

export type SourceType =
  | "manual"
  | "linkedin_manual"
  | "linkedin"
  | "recruiter"
  | "recruiter_email"
  | "referral"
  | "company_site"
  | "job_board"
  | "networking"
  | "direct_outreach"
  | "internal_referral"
  | "greenhouse"
  | "lever"
  | "ashby"
  | "workable"
  | "other";

export interface Company {
  id: string;
  name: string;
  website_url: string | null;
}

export interface JobSource {
  id: string;
  name: string;
  source_type: string | null;
}

export interface Role {
  id: string;
  title: string;
  company_id: string;
  source_id: string | null;
  job_url: string | null;
  location: string | null;
  remote_type: string | null;
  compensation_min: string | null;
  compensation_max: string | null;
  compensation_currency: string | null;
  raw_description: string | null;
  normalized_description: string | null;
  parse_metadata: RoleParseMetadata;
  status: RoleStatus;
  date_found: string;
  date_posted: string | null;
  created_at: string;
  updated_at: string;
  company: Company;
  source: JobSource | null;
}

export interface RoleParseMetadata {
  parserVersion?: string;
  aiModel?: string;
  parseTimestamp?: string;
  parseWarnings?: string[];
  confidence?: Record<string, number>;
  userEditedFields?: string[];
  extractedSkills?: string[];
  opportunityIntelligence?: OpportunityIntelligence;
  [key: string]: unknown;
}

export interface OpportunitySignal {
  type: string;
  label: string;
  severity: "low" | "medium" | "high" | string;
  reason: string;
  basis: string;
  confidence: string;
  evidence: unknown[];
}

export interface OpportunityIntelligence {
  version: string;
  evaluatedAt: string;
  signals: OpportunitySignal[];
  categories: string[];
  summary: string;
}

export interface RoleCreatePayload {
  title: string;
  company: {
    name: string;
    website_url?: string | null;
  };
  source: {
    source_type: SourceType;
  };
  job_url?: string | null;
  location?: string | null;
  remote_type?: string | null;
  compensation_min?: string | null;
  compensation_max?: string | null;
  compensation_currency?: string | null;
  raw_description?: string | null;
  normalized_description?: string | null;
  parse_metadata?: RoleParseMetadata;
  status: RoleStatus;
  date_found?: string | null;
  date_posted?: string | null;
}

export interface RoleParseRequest {
  rawText: string;
  source?: SourceType | null;
  jobUrl?: string | null;
}

export interface ParsedRole {
  roleTitle?: string | null;
  company?: string | null;
  companyWebsite?: string | null;
  jobUrl?: string | null;
  source?: SourceType | null;
  location?: string | null;
  remoteType?: string | null;
  compensationMin?: number | string | null;
  compensationMax?: number | string | null;
  currency?: string | null;
  employmentType?: string | null;
  seniority?: string | null;
  datePosted?: string | null;
  normalizedDescription?: string | null;
  extractedSkills?: string[];
  warnings?: string[];
  confidence?: Record<string, number>;
}

export interface RoleParseResponse {
  parsed: ParsedRole;
  metadata: {
    parserVersion: string;
    model: string;
  };
}

export interface RoleUpdatePayload {
  title?: string;
  status?: RoleStatus;
  remote_type?: string | null;
  normalized_description?: string | null;
}

export type Opportunity = Role;
export type OpportunityCreatePayload = RoleCreatePayload;
export type OpportunityUpdatePayload = RoleUpdatePayload;
export type OpportunityParseRequest = RoleParseRequest;
export type OpportunityParseResponse = RoleParseResponse;
